In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [ ]:
df = pd.read_csv("/content/drive/MyDrive/DataSets/titanic.csv")

In [ ]:
display(df.head())

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df = df.dropna(subset = "Age")

In [ ]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          529
Embarked         2
dtype: int64

#Decision Tree Classifier

In [ ]:
X = df[["PassengerId", "Pclass", "Age", "SibSp", "Parch", "Fare"]]
y = df["Survived"]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, train_size = 0.75)
print("The length of the initial dataset is :", len(X))
print("The length of the train dataset is   :", len(X_train))
print("The length of the test dataset is    :", len(X_test))

The length of the initial dataset is : 714
The length of the train dataset is   : 535
The length of the test dataset is    : 179


In [ ]:
from sklearn.tree import DecisionTreeClassifier

modelDTC = DecisionTreeClassifier()
modelDTC.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
from sklearn.metrics import accuracy_score

x_pred = modelDTC.predict(X_train)
y_pred = modelDTC.predict(X_test)

accuracy1 = accuracy_score(y_train, x_pred)
accuracy2 = accuracy_score(y_test, y_pred)
print(f"Accuracy on train: {round(accuracy1, 2)}")
print(f"Accuracy on test: {round(accuracy2, 2)}")

Accuracy on train: 1.0
Accuracy on test: 0.67


#GridSearch

In [ ]:
from sklearn.model_selection import GridSearchCV
dico = {"max_depth": range(1, 51), "min_samples_leaf" : range(1, 16), "min_samples_split" : [2, 5, 7, 10, 15, 30]}
grid = GridSearchCV(DecisionTreeClassifier(), dico )
grid.fit(X,y)

GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': range(1, 51),
                         'min_samples_leaf': range(1, 16),
                         'min_samples_split': [2, 5, 7, 10, 15, 30]})

In [ ]:
print("best score:",grid.best_score_)
print("best parameters:",grid.best_params_)

best score: 0.710154634098296
best parameters: {'max_depth': 4, 'min_samples_leaf': 3, 'min_samples_split': 2}


In [ ]:
grid.cv_results_

{'mean_fit_time': array([0.00292921, 0.00299392, 0.00241714, ..., 0.00605888, 0.00539637,
        0.00559278]),
 'std_fit_time': array([5.23765909e-04, 3.98375390e-04, 2.56365502e-05, ...,
        1.05207109e-03, 1.97395207e-04, 3.26646171e-04]),
 'mean_score_time': array([0.00188909, 0.00208397, 0.0017725 , ..., 0.00342159, 0.00349078,
        0.00348783]),
 'std_score_time': array([1.17385803e-04, 3.05772414e-04, 2.73313808e-05, ...,
        1.34543252e-04, 7.23933351e-05, 9.56113795e-05]),
 'param_max_depth': masked_array(data=[1, 1, 1, ..., 50, 50, 50],
              mask=[False, False, False, ..., False, False, False],
        fill_value='?',
             dtype=object),
 'param_min_samples_leaf': masked_array(data=[1, 1, 1, ..., 15, 15, 15],
              mask=[False, False, False, ..., False, False, False],
        fill_value='?',
             dtype=object),
 'param_min_samples_split': masked_array(data=[2, 5, 7, ..., 10, 15, 30],
              mask=[False, False, False, ..., Fal

#RandomSearch

In [ ]:
from sklearn.model_selection import RandomizedSearchCV
dico = {"max_depth": range(1, 51), "min_samples_leaf" : range(1, 16), "min_samples_split" : [2, 5, 7, 10, 15, 30]}
rando = RandomizedSearchCV(DecisionTreeClassifier(), dico ,
                           n_iter=200)
rando.fit(X,y)

RandomizedSearchCV(estimator=DecisionTreeClassifier(), n_iter=200,
                   param_distributions={'max_depth': range(1, 51),
                                        'min_samples_leaf': range(1, 16),
                                        'min_samples_split': [2, 5, 7, 10, 15,
                                                              30]})

In [ ]:
print("best score:", rando.best_score_)
print("best parameters:", rando.best_params_)

best score: 0.703161627105289
best parameters: {'min_samples_split': 7, 'min_samples_leaf': 6, 'max_depth': 3}


In [ ]:
rando.cv_results_

{'mean_fit_time': array([0.00589051, 0.01017418, 0.00516868, 0.00521908, 0.01305618,
        0.00858831, 0.00549984, 0.00821457, 0.00829191, 0.00530152,
        0.00531416, 0.00791521, 0.00945616, 0.01323051, 0.00549302,
        0.00321569, 0.00324512, 0.00313578, 0.00321331, 0.00304298,
        0.0032505 , 0.00347967, 0.00316291, 0.00379658, 0.00339088,
        0.00317488, 0.00399141, 0.00336676, 0.00334768, 0.00358744,
        0.0025331 , 0.00360436, 0.00322566, 0.00318708, 0.00312724,
        0.00469265, 0.00316658, 0.00316725, 0.00329809, 0.00316839,
        0.00342054, 0.00339918, 0.00357547, 0.00322204, 0.00323224,
        0.00326195, 0.003724  , 0.00323353, 0.00408645, 0.00332198,
        0.00324798, 0.0033277 , 0.00315447, 0.00356441, 0.00364437,
        0.00359788, 0.00338979, 0.00330119, 0.0036695 , 0.00337114,
        0.00351424, 0.00405078, 0.00340867, 0.00378075, 0.00400958,
        0.00357633, 0.00327787, 0.00237145, 0.00344176, 0.00303421,
        0.00495987, 0.00345321,

Ici on constate que les scores de grid et random sont très différents pour min_samples_split (2 pour grid et 7 pour random) et min_samples_leaf (3 pour grid et 6 pour random)

En revanche ils sont proches pour max_depth (2 pour grid, 3 pour random)

On constate également que les scores des deux méthodes sont relativement proches

0.71 pour grid et 0.70 pour random

Dans le cas présent, le fait de tester les paramètres sur tout le dataset ou sur des éléments aléatoires ne fait pas grande différence